### PR TITLE
Quantum state classes

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Adds quantum state classes to `quantum_info` package. These can be evolved by operator class objects, instructions or circuits.

### Details and comments

* Adds `QuantumState` abstract base class
* Adds `Statevector` class for complex statevector. Can be evolved by Operators and circuits containing gate instructions.
* Adds `DensityMatrix` class for density matrices. Can be evolved by Operators, QuantumChannels, and circuits containing gates, reset, and kraus instructions.

### TODO

* [x] Tests for Statevector
* [x] Tests for DensityMatrix
* ~~Integration with partial_trace and state_fidelity functions~~ (*follow up PR*)

### CODE REVIEW

**Comment by chriseclectic** [comment]
@ajavadia ready for review

File: test/python/quantum_info/operators/channel/test_chi.py [hunk 1]
```diff
@@ -77,26 +77,26 @@ def test_evolve(self):
         # Identity channel
         chan = Chi(self.chiI)
         target_rho = np.array([[0, 0], [0, 1]])
-        self.assertAllClose(chan._evolve(input_psi), target_rho)
-        self.assertAllClose(chan._evolve(np.array(input_psi)), target_rho)
-        self.assertAllClose(chan._evolve(input_rho), target_rho)
-        self.assertAllClose(chan._evolve(np.array(input_rho)), target_rho)
+        self.assertAllClose(chan._evolve(input_psi).data, target_rho)
```
Comment (ajavadia) [inline test/python/quantum_info/operators/channel/test_chi.py]: can you avoid using private methods in the tests? There are quite a lot of them. It should be possible to write tests using the public API.

**Comment by chriseclectic** [comment]
@ajavadia `_evolve` is removed from Operators and base class but it is a helper internal method in `QuantumChannels` that states classes use in their implementation of `evolve` for those objects. The `_evovle` method of channels isn't intended to be called by the user directly (thus the underscore).

File: test/python/quantum_info/operators/channel/test_chi.py (hunk 1 above)
Comment (chriseclectic) [reply test/python/quantum_info/operators/channel/test_chi.py]: These could be changed to `DensityMatrix(input_rho).evolve(chan)`

File: test/python/quantum_info/operators/channel/test_evolve.py [hunk 2]
```diff
@@ -41,10 +42,9 @@ def _unitary_to_other(self, rep, qubits_test_cases, repetitions):
             for _ in range(repetitions):
                 rho = self.rand_rho(dim)
```
Comment (ewinston) [inline test/python/quantum_info/operators/channel/test_evolve.py]: There was some discussion about using random in tests. Did we come to an agreement about that?

Comment (chriseclectic) [reply test/python/quantum_info/operators/channel/test_evolve.py]: I'm not aware of the outcome of that discussion, but I can say I found the random tests quite helpful for finding bugs

Comment (kdk) [reply test/python/quantum_info/operators/channel/test_evolve.py]: More or less, the outcome was that tests that depend on randomness should set a seed, use a `RandomState`, and log the seed (see #1863). If you have tests that are intended to run continuously to look for bugs, they should live in `test/randomized`.

**Comment by chriseclectic** [comment]
Updates

* Added `Statevector.from_label` and `DensityMatrix.from_label` class methods that construct from a label of Pauli eigenstates `0,1,+,-,r,l`.

* Removed `__init__` from instructions or circuits from class init and added `.from_instruction` class method that can initialize a Statevector or Density matrix from an instruction/circuit assuming the initial state is the all zero state (different initial states and be achieved the *standard* way of `Statevector(init_state).evolve(circuit)`).

File: test/python/quantum_info/operators/channel/test_evolve.py (hunk 2 above)
Comment (chriseclectic) [reply test/python/quantum_info/operators/channel/test_evolve.py]: Added logging as per example in 1863

**Comment by chriseclectic** [comment]
@ajavadia @ewinston made the changes, should be good to go now

**Pull request merged**